packages/util/src/primatives.ts

```diff
@@ -0,0 +1,208 @@
+  Pick<
+    SVGElementTagNameMap,
+    Exclude<keyof SVGElementTagNameMap, keyof HTMLElementTagNameMap>
+  >
+
+export type PropsWithAs<CompType extends As, ComponentProps> = ComponentProps &
+  Omit<ComponentPropsWithRef<CompType>, 'as' | keyof ComponentProps> & {
+    as?: CompType
+  }
+
+export type PropsFromAs<CompType extends As, ComponentProps> = (PropsWithAs<
+  CompType,
+  ComponentProps
+> & { as: CompType }) &
+  PropsWithAs<CompType, ComponentProps>
+
+export interface FunctionComponentWithAs<
+  DefaultComponentType extends As,
+  ComponentProps
+> {
+  /**
+   * Inherited from React.FunctionComponent with modifications to support `as`
+   */
+  <CompType extends As>(
+    props: PropsWithAs<CompType, ComponentProps>,
+    context?: any
+  ): ReactElement<any, any> | null
+  (
+    props: PropsWithAs<DefaultComponentType, ComponentProps>,
+    context?: any
+  ): ReactElement<any, any> | null
+
+  /**
+   * Inherited from React.FunctionComponent
+   */
+  displayName?: string
+  propTypes?: WeakValidationMap<
+    PropsWithAs<DefaultComponentType, ComponentProps>
+  >
+  contextTypes?: ValidationMap<any>
+  defaultProps?: Partial<PropsWithAs<DefaultComponentType, ComponentProps>>
+}
+
+interface ExoticComponentWithAs<
+  DefaultComponentType extends As,
+  ComponentProps
+> {
+  /**
+   * **NOTE**: Exotic components are not callable.
+   * Inherited from React.ExoticComponent with modifications to support `as`
+   */
+  (
+    props: PropsWithAs<DefaultComponentType, ComponentProps>
+  ): ReactElement | null
+  <CompType extends As>(
+    props: PropsWithAs<CompType, ComponentProps> & {
+      as: CompType
+    }
+  ): ReactElement | null
+
+  /**
+   * Inherited from React.ExoticComponent
+   */
+  readonly $$typeof: symbol
+}
+
+interface NamedExoticComponentWithAs<
+  DefaultComponentType extends As,
+  ComponentProps
+> extends ExoticComponentWithAs<DefaultComponentType, ComponentProps> {
+  /**
+   * Inherited from React.NamedExoticComponent
+   */
+  displayName?: string
+}
+
+export interface ForwardRefExoticComponentWithAs<
+  DefaultComponentType extends As,
+  ComponentProps
+> extends NamedExoticComponentWithAs<DefaultComponentType, ComponentProps> {
+  /**
+   * Inherited from React.ForwardRefExoticComponent
+   * Will show `ForwardRef(${Component.displayName || Component.name})` in devtools by default,
+   * but can be given its own specific name
+   */
+  defaultProps?: Partial<PropsWithAs<DefaultComponentType, ComponentProps>>
+  propTypes?: WeakValidationMap<
+    PropsWithAs<DefaultComponentType, ComponentProps>
+  >
+}
+
+export interface MemoExoticComponentWithAs<
+  DefaultComponentType extends As,
+  ComponentProps
+> extends NamedExoticComponentWithAs<DefaultComponentType, ComponentProps> {
+  readonly type: DefaultComponentType extends ComponentType
+    ? DefaultComponentType
+    : FunctionComponentWithAs<DefaultComponentType, ComponentProps>
+}
+
+export interface ForwardRefWithAsRenderFunction<
```

Comment:
oops. forgot to document using a function render props. i'll add an example - the approach is pretty straightforward

interface Props {
  children?: React.ReactNode | (() => JSX.Element);
}
const TestInput = forwardRefWithAs<Props, "input">((props, ref) => {
  const { as: Comp = "input", children, ...rest } = props;

  return (
    <Comp ref={ref} {...rest}>
      {isFunction(children) ? children() : children}
    </Comp>
  );
});
```